### PR TITLE
Fix to node exceed limit of 100

### PR DESCRIPTION
**Pull request by IsaacHardy**
### What does this PR do, or why is it needed?
Fixes an issue where filtering to many content items caused the query to break.

### What design trade-offs/decisions were made?
Relevant link for determining node limit calculations: https://github.com/OData/RESTier/issues/579#issuecomment-326976015

## TODO:

- [ ] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [ ] Closes [tag-issues-here]
- [ ] No new warnings in tests, in storybook, and in-app
- [ ] Upload GIF(s) of iOS and Android
- [ ] Set a relevant reviewer

## REVIEW:

**Manual QA**

- [ ] Manual QA on iOS and ensure it looks/behaves as expected
- [ ] Manual QA on Android and ensure it looks/behaves as expected

**Code Review: Questions to consider**

- [ ] Read through the "Files changed" tab _very carefully_
- [ ] Edge cases: what assumptions are made about input?
- [ ] What kind of tests could be written?
- [ ] How might we make this easier for someone else to understand?
- [ ] Could the code be simpler?
- [ ] Will the code be easy to modify in the future?
- [ ] What's one part of these changes that makes you excited to merge it?

> The purpose of PR Review is to _improve the quality of the software._

### CODE REVIEW

File: packages/apollos-data-connector-rock/src/content-items/data-source.js [hunk 1]
```diff
@@ -179,21 +179,36 @@ export default class ContentItem extends RockApolloDataSource {
 
   expanded = true;
 
+  splitFilter = ({ filter }) => {
```
Comment (IsaacHardy) [inline packages/apollos-data-connector-rock/src/content-items/data-source.js]: looking at this a second time, it seems `splitRequest` might make more sense.

File: packages/apollos-rock-apollo-data-source/src/index.js [hunk 2]
```diff
@@ -71,18 +71,42 @@ export default class RockApolloDataSource extends RESTDataSource {
       }
     }
 
+    if (!cursor) return [];
+    try {
+      if (Array.isArray(cursor)) {
+        const edges = await cursor.map((set) =>
```
[the remark this answers is not in the document]
Comment (IsaacHardy) [reply packages/apollos-rock-apollo-data-source/src/index.js]: It is wrapped in a try/catch. Or are you talking about something else?

File: packages/apollos-data-connector-rock/src/content-items/data-source.js [hunk 3]
```diff
@@ -179,21 +179,36 @@ export default class ContentItem extends RockApolloDataSource {
...
+    const requests = [];
+
+    const associationsFiltered = chunk(filter, 15);
+    // eslint-disable-next-line
+    for (let i = 0; i < associationsFiltered.length; i++) {
```
[the remark this answers is not in the document]
Comment (IsaacHardy) [reply packages/apollos-data-connector-rock/src/content-items/data-source.js]: `map` does have a `index` argument so I could. Is that your preference?

File: packages/apollos-rock-apollo-data-source/src/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinnyjth) [reply packages/apollos-rock-apollo-data-source/src/index.js]: Sorry, that wasn't clear :) Move the try/catch to cover the whole method.

File: packages/apollos-data-connector-rock/src/content-items/data-source.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (vinnyjth) [reply packages/apollos-data-connector-rock/src/content-items/data-source.js]: Totally.

File: packages/apollos-rock-apollo-data-source/src/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (IsaacHardy) [reply packages/apollos-rock-apollo-data-source/src/index.js]: I was hesitant to do that since like 70 is already handling the error potential. I could cover the below single cursor edge case within the try/catch, though.

**Comment by IsaacHardy** [comment]
> It's very important that the end user of the request-builder should not be aware of needing the node limit

How would a user current be aware of the node limit with my current implementation?

> My initial idea to take this PR closer to the design goals would be to move the splitFilter/splitRequest function inside the requestBuilder and hide the implementation of it inside get. When get is called, we should execute the request in multiple steps if needed, and then merge the results.

I didn't go this route because the `path` from within the `get` method is already completed with timestamp in all. We would have to disect the entire `path` creation to keep the "logic" within the `get` method. Potentially better, we would have to place that logic within the path method inside the request-builder.

Instead, I decided to extend the Rock Content Item datasource (since this is a ROCK issue, first) and not a request builder issue. I definitely see the value in having some type of mechanism built within the core structure that helps deal with these large queries, however.

**Comment by vinnyjth** [comment]
> How would a user current be aware of the node limit with my current implementation?

They have to make a choice to call `splitRequest` instead of using the default `this.request` tools.

> Instead, I decided to extend the Rock Content Item datasource (since this is a ROCK issue, first) and not a request builder issue. 

I think we're okay putting logic specific to Rock instead of RequestBuilder :) over time it's grown to become a pretty much Rock specific request builder. Look at a lot of the `filter` methods, that use OData to build queries. 

>I definitely see the value in having some type of mechanism built within the core structure that helps deal with these large queries, however.

Yah, I think this is the most important point. Another way to rephrase my goals from above is that whatever solution we come up with, it shouldn't require any changes to the existing `DataSources`. (We might need to change the base `RockDataSource`, but it shouldn't change the ContentItems's data source, for example). The fact we are splitting queries should be entirely invisible to the user, even if that means we end up making more or inefficient requests to Rock. 

I would be happy to brainstorm on this more. My idea around building the query splitting into the `get` method is just an idea, and I would be happy to accept any solution that meets the goals above :)

**Comment by conrad-vanl** [comment]
Closing in favor of #855
